### PR TITLE
Feat:  unsaved file content

### PR DESCRIPTION
**Pull request by tofarr**
### Before Changes

When you make changes in the code editor, if you click into a different file, the changes are lost.
e.g.:
I write some changes:
image

But if I click into a different file, they are lost! (There is no dialog or warning to prevent this!)
image

### After Changes

We store changes even after you click out of a file. Files with unsaved changes are marked with a "*", and a cancel button was added to clear changes. (My original idea was to use local storage / indexed db to store changes, but I have not implemented this for this iteration - so a page refresh will still clear edits)
image

I can click to a different file and make changes there too:
image

Clicking back to the original displays the unsaved changes:
image

Changes are only saved / cancelled when the page is refreshed or the button is clicked.

### Under the hood

I added an array of fileStates to the redux state - this allows for determining whether there is unsavedContent, and displaying different UI states in the file explorer

### Open Questions

I think some UI review is in order - making sure this is consistent with where we want the product to go, and that this UI is intuitive.

### CODE REVIEW

**Comment by amanape** [comment]
Sorry for the delay, I'll give it a full review tomorrow! LGTM though from a quick skim if you're in a rush

File: frontend/src/components/file-explorer/CodeEditor.tsx [hunk 1]
```diff
@@ -84,12 +96,19 @@ function CodeEditor(): JSX.Element {
...
       setShowSaveNotification(true);
-      setLastSavedContent(code);
-      setHasUnsavedChanges(false);
       setTimeout(() => setShowSaveNotification(false), 2000);
+      const newFileState = {
+        path: activeFilepath,
+        savedContent: newContent,
+        unsavedContent: newContent,
+      };
+      setTimeout(() => dispatch(addOrUpdateFileState(newFileState)), 2000);
```
Comment (amanape) [inline frontend/src/components/file-explorer/CodeEditor.tsx:111]: We should consider moving this into a `useEffect` that triggers when `saveStatus` changes. That way, we can also safely clear the timeout for the cases the user navigates away from the code editor.

What do you think?

**Review by amanape — APPROVED** [review]
LGTM! Consider extending the test cases to serve as another form of documentation for newcomers (if you have the time of course!)

File: frontend/src/components/file-explorer/CodeEditor.tsx (hunk 1 above)
Comment (tofarr) [reply frontend/src/components/file-explorer/CodeEditor.tsx:111]: Good idea - gonna implement that now.

**Comment by tofarr** [comment]
Added a few more minimal test cases. TBH I think we need more test cases around editing

**Pull request merged**